### PR TITLE
Create forgot password api endpoints

### PR DESCRIPTION
**Pull request by DwinaTech**
Added two column to user table and model, created forgot endpoint first will check if user is exist then will update resetPasswordExpires and resetPasswordToken then will send email to user the email include the link to allow user to go to reset form.

At reset form will check if token is valid will render the form else will display invalid token message then will submit the data will check again if token is valid then with check if the two passwords user insert them is match if true will send email to user, the email includes the password changes message.

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -8,15 +8,15 @@
...
-      "git pull; pm2 kill; pm2 start --interpreter babel-node bin/www; pm2 logs",
+    "start-prod": "git pull; pm2 kill; pm2 start --interpreter babel-node bin/www; pm2 logs",
     "lint": "./node_modules/.bin/eslint **/*.js",
     "fix": "find ./node_modules/* -mtime +10950 -exec touch {} \\;"
   },
   "engines": {
     "node": ">=8.0.0"
   },
   "dependencies": {
+    "async": "^2.6.1",
```
Comment (kkarimi) [inline backend/package.json]: This library is not necessary

File: backend/src/routes/users.js [hunk 2]
```diff
@@ -148,4 +155,122 @@ router.get('/users/profile', passport.authenticate('jwt', { session: false }), (
   res.json({ success: true })
 })
 
+router.post('/forgot', (req, res) => {
+  async.waterfall([
```
Comment (kkarimi) [inline backend/src/routes/users.js]: This is exactly why we can use `await` for, no need for external library. Await waits until that line is done. For other cases you can use `Promise.all()`

File: backend/src/routes/users.js [hunk 3]
```diff
@@ -148,4 +155,122 @@ router.get('/users/profile', passport.authenticate('jwt', { session: false }), (
...
+            resetPasswordToken = token;
+            resetPasswordExpires = (Date.now() + 3600000).toString(); // 1 hour
+            updateUser(user[0].id, {
+              resetPasswordExpires, resetPasswordToken
+            })
+              .then(userInfo => {
+                if (!userInfo || userInfo.length <= 0) {
+                  res.status(502).json({ success: false, message: 'There is an error occurred!' });
+                } else {
+                  nodemailerMailgun.sendMail({
```
Comment (kkarimi) [inline backend/src/routes/users.js]: Let's move this to a separate function and keep no function or handler longer than 10 lines

File: backend/src/routes/users.js [hunk 4]
```diff
@@ -148,4 +155,122 @@ router.get('/users/profile', passport.authenticate('jwt', { session: false }), (
...
+    if (error) {
+      res.status(502).json({ success: false, error })
+    } else {
+      res.status(200).json({ success: true, data })
+    }
+  });
+});
+
+router.get('/reset/:token', (req, res) => {
+  validateResetInfo(req.params.token, Date.now().toString())
```
Comment (kkarimi) [inline backend/src/routes/users.js]: try/catch, await, no `.then()`

File: backend/src/routes/users.js [hunk 5]
```diff
@@ -148,4 +155,122 @@ router.get('/users/profile', passport.authenticate('jwt', { session: false }), (
...
+      } else {
+        const userId = user[0].id;
+        const { email } = user[0];
+        res.status(200).json({ success: true, userId, email })
+      }
+    })
+});
+
+router.post('/reset/:token', (req, res) => {
+  async.waterfall([
```
Comment (kkarimi) [inline backend/src/routes/users.js]: ditto

File: backend/src/routes/users.js (hunk 2 above)
Comment (DwinaTech) [reply backend/src/routes/users.js]: Done

File: backend/package.json (hunk 1 above)
Comment (DwinaTech) [reply backend/package.json]: Done

File: backend/src/routes/users.js (hunk 3 above)
Comment (DwinaTech) [reply backend/src/routes/users.js]: Done

File: backend/src/routes/users.js (hunk 4 above)
Comment (DwinaTech) [reply backend/src/routes/users.js]: Done

File: backend/src/routes/users.js (hunk 5 above)
Comment (DwinaTech) [reply backend/src/routes/users.js]: Done

File: backend/src/routes/users.js [hunk 6]
```diff
@@ -148,4 +154,79 @@ router.get('/users/profile', passport.authenticate('jwt', { session: false }), (
...
+  crypto.randomBytes(20, async (err, buf) => {
+    if (err) {
+      res.status(502).json({ success: false, message: 'an error occurred!', err })
+    }
+    try {
+      const token = buf.toString('hex');
+      const user = await getUserByEmail(req.body.email);
+      let { resetPasswordExpires, resetPasswordToken } = user[0];
+      resetPasswordToken = token;
+      resetPasswordExpires = (Date.now() + 3600000).toString(); // 1 hour
```
Comment (LouiaR) [inline backend/src/routes/users.js]: You can use moment library to handle time and date

Comment (DwinaTech) [reply backend/src/routes/users.js]: I do not think so good idea to use whole library for just this line if we are using date many time might be helpful

File: backend/src/routes/users.js [hunk 7]
```diff
@@ -148,4 +154,79 @@ router.get('/users/profile', passport.authenticate('jwt', { session: false }), (
...
+
+      await updateUser(user[0].id, {
+        resetPasswordExpires, resetPasswordToken
+      })
+
+      await nodemailerMailgun.sendMail({
+        from: `${req.body.sideEmail}`,
```
Comment (kkarimi) [inline backend/src/routes/users.js]: what is `side`?

File: backend/src/routes/users.js [hunk 8]
```diff
@@ -11,10 +14,13 @@ import {
   updateUser,
   deleteUser,
   getUserByEmail,
-  comparePassword
+  comparePassword,
+  validateResetInfo
 } from '../controllers/users_controller';
 
 const router = express.Router();
+const auth = { auth: { api_key: process.env.API_KEY, domain: process.env.DOMAIN } }
```
Comment (kkarimi) [inline backend/src/routes/users.js]: `API_KEY` is so generic and not explicit, I am not sure what service provided this API?

File: backend/src/routes/users.js (hunk 7 above)
Comment (DwinaTech) [reply backend/src/routes/users.js]: sorry I was mean siteEmail, it is corrected now.

File: backend/src/routes/users.js (hunk 8 above)
Comment (DwinaTech) [reply backend/src/routes/users.js]: Updated to more understandable name

**Comment by kkarimi** [comment]
@DwinaTech has conflicts

**Pull request merged**